### PR TITLE
feat: CRUD functions for Task and Sprint

### LINKED ISSUE #11: CRUD functions for Task model
**Requirements**
We need an internal API for reading and writing Task records.

**Acceptance Criteria**

1. Read functions `getTask(id: number)`, `getTask(params: any)` and `getTasks(params: any)` for searching for task(s)
2. Write functions `createTask(id: number)`, `updateTask(id: number)`, `deleteTask(id: number)`
3. House all Task CRUD functions in `/lib/db/task.ts`.
4. Write unit tests in `/tests/db/task.test.ts`.
 
Check out [TypeScript Unit Testing 101](https://www.testim.io/blog/typescript-unit-testing-101/) for examples.

### LINKED ISSUE #12: CRUD functions for Sprint model
**Requirements**
We need an internal API for reading and writing Sprint records.

**Acceptance Criteria**

1. Read functions `getSprint(params)` and `getSprints(params)` for retrieving sprint(s)
2. Write functions `createSprint(id: number)`, `updateSprint(id: number)`, `deleteSprint(id: number)`
3. House all Sprint CRUD functions in `/lib/db/sprint.ts`.
4. Write unit tests in `/tests/sprint.test.ts`. (Check out [TypeScript Unit Testing 101](https://www.testim.io/blog/typescript-unit-testing-101/) for examples.)

### PR DESCRIPTION
**Pull request by winstuhn**
Changes:

1. Added Unit Tests for Task and Sprint CRUD functionality
2. Updated schema
- removed mapping between sprints and status
- new fields to Project and Task
3. Wrote CRUD functionality for Task and Sprint
4. Wrote Create and Delete functionality for Priority, Project, and Status
5. Create custom Type for Task and Sprint
- Extends GetPayload to include mapped relational fields.

### CODE REVIEW

File: prisma/dev.db
Comment (wasimxyz) [inline prisma/dev.db:1]: I see some data got added to the Project, Sprint, and Task tables. Could you make sure all test-generated data gets cleaned out after a test run?

File: lib/tests/sprint.test.ts
Comment (wasimxyz) [inline lib/tests/sprint.test.ts:1]: I was unable to run this test on my end. Here's the output:

```
wasim@Wasims-MacBook-Pro tests % jest
 FAIL  ./task.test.ts
  ● Test suite failed to run

    ../db/sprint.ts:72:5 - error TS2322: Type '{ project: { connect: { id: any; }; }; title: any; start_date: any; end_date: any; planned_capacity: any; }' is not assignable to type '(Without<SprintCreateInput, SprintUncheckedCreateInput> & SprintUncheckedCreateInput) | (Without<...> & SprintCreateInput)'.
      Type '{ project: { connect: { id: any; }; }; title: any; start_date: any; end_date: any; planned_capacity: any; }' is not assignable to type 'Without<SprintUncheckedCreateInput, SprintCreateInput> & SprintCreateInput'.
        Property 'status' is missing in type '{ project: { connect: { id: any; }; }; title: any; start_date: any; end_date: any; planned_capacity: any; }' but required in type 'SprintCreateInput'.

    72     data: {
           ~~~~

      ../../node_modules/.prisma/client/index.d.ts:11496:5
        11496     status: StatusCreateNestedOneWithoutSprintsInput
                  ~~~~~~
        'status' is declared here.
      ../../node_modules/.prisma/client/index.d.ts:5606:5
        5606     data: XOR<SprintCreateInput, SprintUncheckedCreateInput>
                 ~~~~
        The expected type comes from property 'data' which is declared here on type '{ select?: SprintSelect<DefaultArgs> | null | undefined; include?: SprintInclude<DefaultArgs> | null | undefined; data: (Without<...> & SprintUncheckedCreateInput) | (Without<...> & SprintCreateInput); }'

 FAIL  ./sprint.test.ts
  ● Test suite failed to run

    ../db/sprint.ts:72:5 - error TS2322: Type '{ project: { connect: { id: any; }; }; title: any; start_date: any; end_date: any; planned_capacity: any; }' is not assignable to type '(Without<SprintCreateInput, SprintUncheckedCreateInput> & SprintUncheckedCreateInput) | (Without<...> & SprintCreateInput)'.
      Type '{ project: { connect: { id: any; }; }; title: any; start_date: any; end_date: any; planned_capacity: any; }' is not assignable to type 'Without<SprintUncheckedCreateInput, SprintCreateInput> & SprintCreateInput'.
        Property 'status' is missing in type '{ project: { connect: { id: any; }; }; title: any; start_date: any; end_date: any; planned_capacity: any; }' but required in type 'SprintCreateInput'.

    72     data: {
           ~~~~

      ../../node_modules/.prisma/client/index.d.ts:11496:5
        11496     status: StatusCreateNestedOneWithoutSprintsInput
                  ~~~~~~
        'status' is declared here.
      ../../node_modules/.prisma/client/index.d.ts:5606:5
        5606     data: XOR<SprintCreateInput, SprintUncheckedCreateInput>
                 ~~~~
        The expected type comes from property 'data' which is declared here on type '{ select?: SprintSelect<DefaultArgs> | null | undefined; include?: SprintInclude<DefaultArgs> | null | undefined; data: (Without<...> & SprintUncheckedCreateInput) | (Without<...> & SprintCreateInput); }'

Test Suites: 2 failed, 2 total
Tests:       0 total
Snapshots:   0 total
Time:        2.052 s
Ran all test suites.
```

File: lib/db/priority.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * Creates a new priority.
+ *
+ * @export
+ * @async
+ * @param {Object} params Information about the priority to create.
+ * @param {string} params.name The name of the priority.
+ * @param {string} params.description The description of the priority.
+ * @returns {Promise<Object>} The created priority object.
+ */
+export async function createPriority(params: any) {
```
Comment (wasimxyz) [inline lib/db/priority.ts]: For other `create` functions, having the doc tag is a big help for anyone using this function, but would make it hard to parse which fields are required and which are not.

Instead, we can just create the object and pass it to this function. The object will have type safety, so if someone tries to write `const priority: Priority = { name: "Medium" }`, they will get an error that the `description` field is missing.

Since this function requires less than 3 fields, it might also be easier for you to create the object for us if you wanted (up to you). Something like `function createPriority(name: string, description: string)` would be appropriate.

File: lib/db/priority.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Deletes a priority.
+ *
+ * @export
+ * @async
+ * @param {number} id The ID of the priority to delete.
+ * @returns {Promise<Object>} The deleted priority object.
+ */
+export async function deletePriority(id: any) {
```
Comment (wasimxyz) [inline lib/db/priority.ts]: Allowing `any` for known types is bad practice. We should use the known type wherever possible :)

File: lib/db/sprint.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+ * @param {number} params.id The ID of the sprint to retrieve.
+ * @param {number} params.project_id The ID of the project the sprint belongs to.
+ * @param {string} params.title The title of the sprint.
+ * @param {Date} params.start_date The start date of the sprint.
+ * @param {Date} params.end_date The end date of the sprint.
+ * @returns {Promise<Object>} The retrieved sprint object.
+ */
+export async function getSprint(params: any): Promise<Object | null> {
+  const sprint = await prisma.sprint.findFirst({
+    where: {
```
Comment (wasimxyz) [inline lib/db/sprint.ts:17]: Instead of manually typing out the parameters in your `where` clause, you can use the [spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax)! That way we don't have to edit this function every time we make a change to the model. Example:

```javascript
where:  { ...params }
```

File: lib/db/sprint.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+ * @async
+ * @param {Object} params Information about the sprint.
+ * @param {number} params.id The ID of the sprint to retrieve.
+ * @param {number} params.project_id The ID of the project the sprint belongs to.
+ * @param {string} params.title The title of the sprint.
+ * @param {Date} params.start_date The start date of the sprint.
+ * @param {Date} params.end_date The end date of the sprint.
+ * @returns {Promise<Object>} The retrieved sprint object.
+ */
+export async function getSprint(params: any): Promise<Object | null> {
```
Comment (wasimxyz) [inline lib/db/sprint.ts]: For type safety, we want the return type for any `get` function to be the model type.

In this case, since you are including other fields, you'll have to use the Prisma payload function to create a custom `Sprint` type as well.

File: lib/db/sprint.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+ * @param {Date} params.start_date The updated start date of the sprint.
+ * @param {Date} params.end_date The updated end date of the sprint.
+ * @param {number} params.planned_capacity The updated planned capacity of the sprint.
+ * @returns {Promise<Object>} The updated sprint object.
+ */
+export async function updateSprint(params: any) {
+  const sprint = await prisma.sprint.update({
+    where: { id: params.id },
+    data: {
+      title: params.title || undefined,
```
Comment (wasimxyz) [inline lib/db/sprint.ts]: The spread syntax would be appropriate to use for `update` functions as well :)

File: prisma/schema.prisma [hunk 6]
```diff
@@ -44,26 +46,25 @@ model Activity {
...
-  assigned_to         Int
-  sprints             Sprint   @relation(fields: [sprint_id], references: [id])
+  description         String?
+  category            String?    
+  requirements        String?
+  acceptance_criteria String?
+  points              Int?
+  user                User?     @relation(fields: [assigned_to], references: [id])    
+  assigned_to         Int?
+  sprint              Sprint    @relation(fields: [sprint_id], references: [id])
```
Comment (wasimxyz) [inline prisma/schema.prisma:68]: Thanks for updating these to optional fields! I was just about to ask for that yesterday

File: lib/db/sprint.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+ * @async
+ * @param {Object} params Information about the sprints to retrieve.
+ * @param {number} params.id The ID of the sprint(s) to retrieve.
+ * @param {number} params.project_id The ID of the project the sprint(s) belong to.
+ * @param {string} params.title The title of the sprint(s).
+ * @param {Date} params.start_date The start date of the sprint(s).
+ * @param {Date} params.end_date The end date of the sprint(s).
+ * @returns {Promise<Object[]>} An array of retrieved sprint objects.
+ */
+export async function getSprints(params: any): Promise<Object[]> {
```
Comment (wasimxyz) [inline lib/db/sprint.ts]: For these get (search) functions, it's best practice to use partial types instead of `any`:

```javascript
getSprints(params: Partial<Sprint>): Promise<Sprint[]> {
   ...
}
```

Comment (wasimxyz) [reply lib/db/sprint.ts]: Additionally, thank you for creating the very descriptive doc tags! I should have given you a heads up about some of the type features in TypeScript before you got started.

Because the types prevent any users of your functions from passing incorrect parameters, you don't have to define the Sprint parameters in the doctags. Additionally, they would be hard to maintain as we make changes to the model in the future.

**Review by wasimxyz — CHANGES_REQUESTED** [review]
Great work Winston! This is going to be a huge help for Joel and I. 

This is a strong PR, just needs some changes with regards to type safety. Tests look great too, just couldn't get them to work. Feel free to reach out offline if you have any questions.

File: lib/tests/sprint.test.ts
Comment (winstuhn) [reply lib/tests/sprint.test.ts:1]: It seems like the error is caused because I removed the relational mapping for Status from Tasks.
I updated the DB and it should be fixed!

Comment (wasimxyz) [reply lib/tests/sprint.test.ts:1]: Unfortunately these tests are still failing for me!

File: lib/db/priority.ts
Comment (wasimxyz) [inline lib/db/priority.ts:1]: Looks great!

File: lib/db/project.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ */
+export async function createProject(
+  params: Partial<Project>
+): Promise<Project> {
+  const project = await prisma.project.create({
+    data: {
+      ...params,
+    },
+  });
+  return project;
```
Comment (wasimxyz) [inline lib/db/project.ts:20]: Nice touch returning the created / modified / deleted record. Will be useful for validation.

File: lib/db/sprint.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+ * @returns {Promise<Sprint>} The created sprint object.
+ */
+export async function createSprint(params: Partial<Sprint>): Promise<Sprint> {
+  const sprint = await prisma.sprint.create({
+    data: {
+      project: { connect: { id: params.project_id } },
+      title: params.title,
+      start_date: params.start_date,
+      end_date: params.end_date,
+      planned_capacity: params.planned_capacity,
```
Comment (wasimxyz) [inline lib/db/sprint.ts:60]: Looks like you forgot to use spread syntax here? I do see that you linked the `project_id` manually -- is this necessary, or would passing `project_id` as part of the `Sprint` object suffice to create the relation?

File: lib/db/sprint.ts [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+ * Retrieves a single sprint based on provided parameters.
+ *
+ * @export
+ * @async
+ * @param {Sprint} params
+ * @returns {Promise<Sprint | null>} The retrieved sprint object or null
+ */
+export async function getSprint(
+  params: Partial<Sprint>
+): Promise<Sprint | null> {
```
Comment (wasimxyz) [inline lib/db/sprint.ts]: Because you're including tasks, you'll need to set the return type equal to the custom `Sprint` type you created in `types.ts` here (otherwise the type safety will prevent us from accessing `sprint.task.property`)

File: lib/db/task.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+/**
+ * Retrieves a single task based on provided parameters.
+ *
+ * @export
+ * @async
+ * @param {Partial<Task>} params Information about the task.
+ * @returns {Promise<Task | null>}
+ *} The retrieved task object.
+ */
+export async function getTask(params: Partial<Task>): Promise<Task | null> {
```
Comment (wasimxyz) [inline lib/db/task.ts]: You'll want to use the custom `Task` type you created in `types.ts` here, as well as in `getTasks()`

File: lib/db/task.ts [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+ * Creates a new task.
+ *
+ * @export
+ * @async
+ * @param {Partial<Task>} params Information about the task to create.
+ * @returns {Promise<Task>} The created task object.
+ */
+export async function createTask(params: any): Promise<Task> {
+  const task = await prisma.task.create({
+    data: {
```
Comment (wasimxyz) [inline lib/db/task.ts:71]: Same note as in `sprint.ts`: can spread syntax be used here?

File: lib/db/task.ts [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+/**
+ * Updates an existing task.
+ *
+ * @export
+ * @async
+ * @param {Partial<Task>} params Information about the task to update.
+ * @returns {Promise<Task>} The updated task object.
+ */
+export async function updateTask(params: Partial<Task>): Promise<Task> {
+  const task = await prisma.task.update({
```
Comment (wasimxyz) [inline lib/db/task.ts:96]: We may want to use the custom types for the `update` functions as well...

Additionally, if the spread syntax isn't flexible for the mapped relations, then we might have to create additional functions (such as `moveToNewSprint` or `updateTaskPriority` in a future story...

File: lib/db/types.ts [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import { PrismaClient, Prisma } from "@prisma/client";
+
+const prisma = new PrismaClient();
+
+export type Task = Prisma.TaskGetPayload<{
+  include: {
+    user: true;
+    status: true;
+    priority: true;
```
Comment (wasimxyz) [inline lib/db/types.ts]: Missing `activities` and `comments`?

**Review by wasimxyz — CHANGES_REQUESTED** [review]
Revisions look great! Requesting a few small changes, let me know if you have any thoughts. Once those tests are passing, this PR is good to go.

**Comment by wasimxyz** [comment]
TIL `jest` runs tests in parallel. To run sequentially, use `jest --runInBand`!

Screenshot 2024-05-01 at 5 52 50 PM

File: lib/db/sprint.ts (hunk 10 above)
Comment (wasimxyz) [reply lib/db/sprint.ts]: @winstuhn

File: lib/db/sprint.ts (hunk 9 above)
Comment (wasimxyz) [reply lib/db/sprint.ts:60]: @winstuhn

File: lib/db/task.ts (hunk 11 above)
Comment (wasimxyz) [reply lib/db/task.ts]: @winstuhn

File: lib/db/task.ts (hunk 12 above)
Comment (wasimxyz) [reply lib/db/task.ts:71]: @winstuhn

File: lib/db/task.ts (hunk 13 above)
Comment (wasimxyz) [reply lib/db/task.ts:96]: @winstuhn

File: lib/db/types.ts (hunk 14 above)
Comment (wasimxyz) [reply lib/db/types.ts]: @winstuhn

**Review by wasimxyz — CHANGES_REQUESTED** [review]

File: prisma/dev.db
Comment (winstuhn) [reply prisma/dev.db:1]: Sure! I reset the database and added delete methods to the unit tests.

File: lib/db/priority.ts (hunk 1 above)
Comment (winstuhn) [reply lib/db/priority.ts]: Thanks for the recommendation! For this function and other functions with <2 parameters, I created an object and passed it to the function.

File: lib/db/sprint.ts (hunk 3 above)
Comment (winstuhn) [reply lib/db/sprint.ts:17]: I didn't know about this! Thanks for the information

File: lib/db/priority.ts (hunk 2 above)
Comment (winstuhn) [reply lib/db/priority.ts]: Modified this, and all other functions!

File: lib/db/sprint.ts (hunk 4 above)
Comment (winstuhn) [reply lib/db/sprint.ts]: Created a custom return type for Sprint and Task: TaskWithIncludes and SprintWithIncludes

File: lib/db/sprint.ts (hunk 5 above)
Comment (winstuhn) [reply lib/db/sprint.ts]: Modified update functions to include spread syntax!

File: prisma/schema.prisma (hunk 6 above)
Comment (winstuhn) [reply prisma/schema.prisma:68]: No problem!

File: lib/db/sprint.ts (hunk 7 above)
Comment (winstuhn) [reply lib/db/sprint.ts]: Modified the doc tags. Passed in the Model Types and Custom Return Types instead.

File: lib/db/sprint.ts (hunk 9 above)
Comment (winstuhn) [reply lib/db/sprint.ts:60]: I think it requires manually linking the related entities. As a result, for the create functions I cannot use the spread syntax. I'll keep looking for a fix for this, however.

File: lib/db/sprint.ts (hunk 10 above)
Comment (winstuhn) [reply lib/db/sprint.ts]: Got it, thanks!

File: lib/db/task.ts (hunk 11 above)
Comment (winstuhn) [reply lib/db/task.ts]: Done! Thanks.

File: lib/db/task.ts (hunk 12 above)
Comment (winstuhn) [reply lib/db/task.ts:71]: I'll keep looking for a work-around, for now, I am unable to use spread syntax without it causing errors.

File: lib/db/task.ts (hunk 13 above)
Comment (winstuhn) [reply lib/db/task.ts:96]: Updated all return types to include the custom types.

File: lib/db/types.ts (hunk 14 above)
Comment (winstuhn) [reply lib/db/types.ts]: Added!

**Review by wasimxyz — APPROVED** [review]
@winstuhn Really great job!!! This is going to be a strong launchpad for @joelhenningsen and I.
